tests/Model/LazyBSONArrayTest.php

```diff
@@ -0,0 +1,289 @@
+
+class LazyBSONArrayTest extends TestCase
+{
+    public static function provideTestArray(): Generator
+    {
+        yield 'array' => [
+            new LazyBSONArray([
+                'bar',
+                new LazyBSONDocument(['bar' => 'baz']),
+                new LazyBSONArray([0, 1, 2]),
+            ]),
+        ];
+
+        yield 'packedArray' => [
+            new LazyBSONArray(PackedArray::fromPHP([
+                'bar',
+                ['bar' => 'baz'],
+                [0, 1, 2],
+            ])),
+        ];
+    }
+
+    public function testConstructWithoutArgument(): void
+    {
+        $instance = new LazyBSONArray();
+        $this->assertSame([], iterator_to_array($instance));
+    }
+
+    public function testConstructWithWrongType(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        new LazyBSONArray('foo');
+    }
+
+    public function testConstructWithArrayUsesLiteralValues(): void
+    {
+        $array = new LazyBSONArray([
+            (object) ['bar' => 'baz'],
+            ['bar' => 'baz'],
+            [0, 1, 2],
+        ]);
+
+        $this->assertInstanceOf(stdClass::class, $array[0]);
+        $this->assertIsArray($array[1]);
+        $this->assertIsArray($array[2]);
+    }
+
+    public function testClone(): void
+    {
+        $original = new LazyBSONArray();
+        $original[0] = (object) ['foo' => 'bar'];
+
+        $clone = clone $original;
+        $clone[0]->foo = 'baz';
+
+        self::assertSame('bar', $original[0]->foo);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetGet(LazyBSONArray $array): void
+    {
+        $this->assertSame('bar', $array[0]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetGetAfterUnset(LazyBSONArray $array): void
+    {
+        $this->assertSame('bar', $array[0]);
+        unset($array[0]);
+
+        $this->expectWarning();
+        $this->expectWarningMessage('Undefined offset: 0');
+        $array[0];
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetGetForMissingOffset(LazyBSONArray $array): void
+    {
+        $this->expectWarning();
+        $this->expectWarningMessage('Undefined offset: 4');
+        $array[4];
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetGetForNumericOffset(LazyBSONArray $array): void
+    {
+        $this->assertSame('bar', $array['0']);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetGetForUnsupportedOffset(LazyBSONArray $array): void
+    {
+        $this->expectWarning();
+        $this->expectWarningMessage('Undefined offset: foo');
+        $array['foo'];
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testGetDocument(LazyBSONArray $array): void
+    {
+        $this->assertInstanceOf(LazyBSONDocument::class, $array[1]);
+        $this->assertInstanceOf(LazyBSONDocument::class, $array[1]);
```

Comment:
Are the duplicate assertions here and in `testGetArray()` intentional in order to test the `$read` property? If so, I think this warrants a comment.

Reply:
Not sure how they came to be, but the duplicate assertions are not intentional. I've removed them.